pkg/trace/remoteconfighandler/remote_config_handler.go

```diff
@@ -0,0 +1,179 @@
+		return nil
+	}
+
+	// paths are in the form datadog/<org ID>/<product name>/<config ID>/config
+	r := regexp.MustCompile(`datadog\/\d+\/APM_SAMPLING\/(.+)\/config`)
+
+	return &RemoteConfigHandler{
+		remoteClient:          conf.RemoteSamplingClient,
+		prioritySampler:       prioritySampler,
+		rareSampler:           rareSampler,
+		errorsSampler:         errorsSampler,
+		agentConfig:           conf,
+		remoteConfigPathRegex: r,
+	}
+}
+
+func (h *RemoteConfigHandler) Start() {
+	if h == nil {
+		return
+	}
+
+	h.remoteClient.Start()
+	h.remoteClient.RegisterAPMUpdate(h.onUpdate)
+}
+
+func (h *RemoteConfigHandler) onUpdate(update map[string]state.APMSamplingConfig) {
+	remoteRateUpdates, samplerconfigPayload, err := h.extractUpdatePayloads(update)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if len(remoteRateUpdates) != 0 {
+		log.Debugf("updating remote rates with updates: %v", remoteRateUpdates)
+		h.prioritySampler.UpdateRemoteRates(remoteRateUpdates)
+	}
+
+	if samplerconfigPayload != nil {
+		log.Debugf("updating samplers with remote configuration: %v", samplerconfigPayload)
+		h.updateSamplers(*samplerconfigPayload)
+	}
+}
+
+func (h *RemoteConfigHandler) extractUpdatePayloads(update map[string]state.APMSamplingConfig) ([]sampler.RemoteRateUpdate, *apmsampling.SamplerConfig, error) {
+	var remoteRateUpdates []sampler.RemoteRateUpdate
+	var samplerconfigPayload *apmsampling.SamplerConfig
+	for path, conf := range update {
+		configID, err := h.configPathToID(path)
+		if err != nil {
+			return []sampler.RemoteRateUpdate{}, nil, err
+		}
+
+		switch configID {
+		case dynamicRatesByService, userDefinedRatesByService:
+			var payload apmsampling.APMSampling
+			_, err := payload.UnmarshalMsg(conf.Config)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal APMSampling remote config envPayload: %w", err)
+			}
+			remoteRateUpdates = append(remoteRateUpdates, sampler.RemoteRateUpdate{Version: conf.Metadata.Version, Config: payload})
+		case samplerconfig:
+			samplerconfigPayload = &apmsampling.SamplerConfig{}
+			err = json.Unmarshal(conf.Config, samplerconfigPayload)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal SamplerConfig remote config envPayload: %w", err)
+			}
+		}
+	}
+	return remoteRateUpdates, samplerconfigPayload, nil
+}
+
+func (h *RemoteConfigHandler) configPathToID(path string) (apmsamplingRemoteConfigID, error) {
+	pathMatch := h.remoteConfigPathRegex.FindStringSubmatch(path)
+	if pathMatch == nil {
+		return "", fmt.Errorf("failed to match remote config path %s", path)
+	}
+
+	switch pathMatch[1] {
+	case string(dynamicRatesByService):
+		return dynamicRatesByService, nil
+	case string(userDefinedRatesByService):
+		return userDefinedRatesByService, nil
+	case string(samplerconfig):
+		return samplerconfig, nil
+	default:
+		return "", fmt.Errorf("unexpected remote config ID: %s", pathMatch[1])
+	}
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 Nitpick:** ‏We could probably simplify this a little (my spacing is probably wrong, but you get the general idea). Though it is fine as it is.

```suggestion
        configType := apmsamplingRemoteConfigID(pathMatch[1])
	if (configType == dynamicRatesByService ||
	    configType == userDefinedRatesByService ||
	    configType == samplerconfig) {
	  return configType, nil   
	}
	return "", fmt.Errorf("unexpected remote config ID: %s", pathMatch[1])
```